### PR TITLE
WT-13819 Add live restore file system directory list

### PR DESCRIPTION
**Pull request by zunyiliu**
There are below changes:
1. Add implementation of `__live_restore_fs_directory_list`.
2. Add implementation of `__live_restore_fs_directory_list_single`.
3. Add implementation of `__live_restore_fs_directory_list_free`.
4. Change the params of `__dest_has_tombstone` so it can be called where file handle is not opened.
5. Change the `__live_restore_fs_backing_filename` when layer is source, now the path it returns is `src_path + "/" + relative_path` instead of `src_path + "/" + base_name`.

### CODE REVIEW

File: src/live_restore/live_restore_fs.c [hunk 1]
```diff
@@ -12,45 +12,47 @@
...
+
 /*
  * __live_restore_fs_backing_filename --
- *     Convert a live restore file path (e..g WT_TEST/WiredTiger.wt) to the actual path of the
- *     backing file. This can be the file in the destination directory (which is identical to the
- *     live restore path), or the file in the source directory. The function allocates memory for
- *     the path string and expects the caller to free it.
+ *     Convert a live restore file/directory path (e..g WT_TEST/WiredTiger.wt) to the actual path of
+ *     the backing file/directory. This can be the file in the destination directory (which is
+ *     identical to the live restore path), or the file in the source directory. The function
```
Comment (luke-pearson) [inline src/live_restore/live_restore_fs.c:22]: I think the "live restore path" spoken about in the parenthetical comment "which is identical to the live restore path". Is better referred to as the WiredTiger home path. So "which is identical to the live restore path" -> "which is identical to the wiredtiger home path".

File: src/live_restore/live_restore_fs.c [hunk 2]
```diff
@@ -12,45 +12,47 @@
...
+ *     allocates memory for the path string and expects the caller to free it. If name is an
+ *     absolute path, it will always be in format "/absolute_prefix/dest_home/relative_path",
+ *     otherwise name always begins with dest_home (e..g dest_home/relative_path). The function
```
Comment (luke-pearson) [inline src/live_restore/live_restore_fs.c:25]: otherwise name always begins -> otherwise name is a relative path which always begins with dest_home

**Comment by luke-pearson** [comment]
As a general comment this PR will hugely benefit from unit testing with example path manipulations. I'd say that alone will take at least a day to implement. I understand it is waiting on Andrews PR being merged which we are prioritizing.

File: src/live_restore/live_restore_fs.c [hunk 3]
```diff
@@ -12,45 +12,47 @@
...
+  const char *dest_home, const char *name, char **pathp)
 {
     WT_DECL_RET;
     size_t len;
-    char *buf, *temp_name;
+    char *buf, *filename, *temp_name;
 
-    temp_name = buf = NULL;
+    buf = filename = temp_name = NULL;
+    WT_RET(__wt_strdup(session, name, &temp_name));
```
Comment (luke-pearson) [inline src/live_restore/live_restore_fs.c:37]: Why do we need a `temp_name` variable here?

File: src/live_restore/live_restore_fs.c [hunk 4]
```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
...
+
+    /* Get files from destination. */
+    WT_ERR(__live_restore_fs_backing_filename(
+      &lr_fs->destination, session, lr_fs->destination.home, directory, &path_dest));
+    WT_ERR(lr_fs->os_file_system->fs_directory_list(
+      lr_fs->os_file_system, wt_session, path_dest, prefix, &dirlist_dest, countp));
+
+    for (namep = dirlist_dest; *namep != NULL; namep++)
+        if (namep != NULL && *namep != NULL &&
+          !(strlen(*namep) >= strlen(".deleted") &&
```
Comment (luke-pearson) [inline src/live_restore/live_restore_fs.c:289]: We know that `*namep != NULL` per the loop check so we don't need to check that in the if statement here. I also think the check for a ".deleted" should be simplified and possibly turned into a function like __name_is_tombstone(char *filename)

Comment (ajmorton) [reply src/live_restore/live_restore_fs.c:289]: +1
This boolean is also pretty difficult to parse can we either store the result in a clearly named bool
```
is_marker_file = (namep != NULL && *namep != NULL &&
          !(strlen(*namep) >= strlen(".deleted") && strcmp(*namep + strlen(*namep) - strlen(".deleted"), ".deleted") == 0);
if (!is_marker_file) {
```
Or add a clarifying comment?

File: src/live_restore/live_restore_fs.c [hunk 5]
```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
...
+                goto done;
+        }
+
+    /* Get files from source. */
+    WT_ERR(__live_restore_fs_backing_filename(
+      &lr_fs->source, session, lr_fs->destination.home, directory, &path_src));
+    WT_ERR(lr_fs->os_file_system->fs_directory_list(
+      lr_fs->os_file_system, wt_session, path_src, prefix, &dirlist_src, countp));
+
+    for (namep = dirlist_src; *namep != NULL; namep++) {
```
Comment (ajmorton) [inline src/live_restore/live_restore_fs.c:305]: What is the value of `namep` in this loop? I have a suspicion that it'll return either `WT_SOURCE/foo.wt` or `foo.wt`, and that needs to be converted to `WT_DEST/foo.wt` before we call has_tombstone.

File: src/live_restore/live_restore_fs.c [hunk 6]
```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
...
+    if (dirlist_src != NULL)
+        WT_TRET(__live_restore_fs_directory_list_free(fs, wt_session, dirlist_src, count_src));
+
+    *dirlistp = entries;
+    *countp = count_dest + count_src;
+    if (ret == 0)
+        return (0);
+
+    WT_TRET(__live_restore_fs_directory_list_free(fs, wt_session, entries, *countp));
+    return (ret);
```
Comment (ajmorton) [inline src/live_restore/live_restore_fs.c:336]:
```suggestion
    if (ret != 0)
        WT_TRET(__live_restore_fs_directory_list_free(fs, wt_session, entries, *countp));

    return (ret);
```

File: src/live_restore/live_restore_fs.c [hunk 7]
```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
     return (0);
 }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
```
Comment (ajmorton) [inline src/live_restore/live_restore_fs.c:252]: Is this still necessary? I don't see an unused parameter in here. If we still need it consider `WT_UNUSED` instead of the pragma

**Review by ajmorton — COMMENTED** [review]
Thanks! Couple of comments and questions.

I know we don't have the unit testing framework ready yet, but what sort of testing have you run locally?

**Comment by zunyiliu** [comment]
> Thanks! Couple of comments and questions.
> 
> I know we don't have the unit testing framework ready yet, but what sort of testing have you run locally?

@ajmorton I didn't test edge cases, I pulled your catch2 test branch as it was in review at that time, and ran two cases when `name` is an absolute path or relative path.

File: src/live_restore/live_restore_fs.c [hunk 8]
```diff
@@ -12,45 +12,47 @@
...
          */
-        filename = basename(temp_name);
-        /* +1 for the path separator, +1 for the null terminator. */
-        len = strlen(layer->home) + 1 + strlen(filename) + 1;
+        filename += strlen(dest_home);
+
+        /* +1 for the null terminator. */
+        len = strlen(layer->home) + strlen(filename) + 1;
         WT_ERR(__wt_calloc(session, 1, len, &buf));
-        WT_ERR(__wt_snprintf(buf, len, "%s%s%s", layer->home, __wt_path_separator(), filename));
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/live_restore/live_restore_fs.c:53]: Unless the source dir is ./ which is impossible I think there is a guarantee that filename starts with `/` with how Zunyi has implemented this.

File: src/live_restore/live_restore_fs.c [hunk 9]
```diff
@@ -247,17 +265,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
...
+ *     destination files, exclude any files matching the marker paths. For source files, exclude
+ *     files that are either marked as tombstones or already present in the destination directory.
  */
 static int
-__live_restore_fs_notsup(WT_SESSION *wt_session)
+__live_restore_fs_directory_list_worker(WT_FILE_SYSTEM *fs, WT_SESSION *wt_session,
+  const char *directory, const char *prefix, char ***dirlistp, uint32_t *countp, bool single)
 {
-    WT_RET_MSG((WT_SESSION_IMPL *)wt_session, ENOTSUP, "Unsupported fs operation");
+    WT_DECL_RET;
```
Comment (luke-pearson) [inline src/live_restore/live_restore_fs.c:277]: Given we don't have to declare variables at the start this function could possibly be cleaned up a bit, it is likely not a big deal though.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: src/live_restore/live_restore_fs.c [hunk 10]
```diff
@@ -15,29 +15,46 @@
 static int __live_restore_fs_directory_list_free(
   WT_FILE_SYSTEM *fs, WT_SESSION *wt_session, char **dirlist, uint32_t count);
 
+/*
+ * __live_restore_create_file_path --
+ *     Generate the file path of a layer for a file's name. This file path does not need to exist.
```
Comment (ajmorton) [inline src/live_restore/live_restore_fs.c:20]:
```suggestion
 *     Generate the path of a file or directory in a layer. The file or directory must exist at the root of the layer.
```

File: src/live_restore/live_restore_fs.c [hunk 11]
```diff
@@ -12,45 +12,64 @@
...
+static int __live_restore_fs_directory_list_free(
+  WT_FILE_SYSTEM *fs, WT_SESSION *wt_session, char **dirlist, uint32_t count);
+
+/*
+ * __live_restore_create_file_path --
+ *     Generate the file path of a layer for a file's name. This file path does not need to exist.
+ */
+static int
+__live_restore_create_file_path(
+  WT_SESSION_IMPL *session, WT_LIVE_RESTORE_FS_LAYER *layer, char *name, char **out)
```
Comment (ajmorton) [inline src/live_restore/live_restore_fs.c:24]:
```suggestion
  WT_SESSION_IMPL *session, WT_LIVE_RESTORE_FS_LAYER *layer, char *name, char **filepathp)
```

**Review by ajmorton — APPROVED** [review]
LGTM 📁
